# ID5030 Assignment 2 - Solution

#### By: P Akhil Reddy    
#### Roll Number: CH20B076

In [1]:
import numpy as np
import pandas as pd
import matplotlib as mpl

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import PolynomialFeatures

In [2]:
df = pd.read_excel('Concrete_Data.xls')

In [3]:
df.head()

Cement (component 1)(kg in a m^3 mixture)  \
0                                      540.0   
1                                      540.0   
2                                      332.5   
3                                      332.5   
4                                      198.6   

   Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
0                                                0.0       
1                                                0.0       
2                                              142.5       
3                                              142.5       
4                                              132.4       

   Fly Ash (component 3)(kg in a m^3 mixture)  \
0                                         0.0   
1                                         0.0   
2                                         0.0   
3                                         0.0   
4                                         0.0   

   Water  (component 4)(kg in a m^3 mixture)  \
0                                      162.0   
1                                      162.0   
2                                      228.0   
3                                      228.0   
4                                      192.0   

   Superplasticizer (component 5)(kg in a m^3 mixture)  \
0                                                2.5     
1                                                2.5     
2                                                0.0     
3                                                0.0     
4                                                0.0     

   Coarse Aggregate  (component 6)(kg in a m^3 mixture)  \
0                                             1040.0      
1                                             1055.0      
2                                              932.0      
3                                              932.0      
4                                              978.4      

   Fine Aggregate (component 7)(kg in a m^3 mixture)  Age (day)  \
0                                              676.0         28   
1                                              676.0         28   
2                                              594.0        270   
3                                              594.0        365   
4                                              825.5        360   

   Concrete compressive strength(MPa, megapascals)   
0                                         79.986111  
1                                         61.887366  
2                                         40.269535  
3                                         41.052780  
4                                         44.296075

In [4]:
df.describe()

Cement (component 1)(kg in a m^3 mixture)  \
count                                1030.000000   
mean                                  281.165631   
std                                   104.507142   
min                                   102.000000   
25%                                   192.375000   
50%                                   272.900000   
75%                                   350.000000   
max                                   540.000000   

       Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
count                                        1030.000000       
mean                                           73.895485       
std                                            86.279104       
min                                             0.000000       
25%                                             0.000000       
50%                                            22.000000       
75%                                           142.950000       
max                                           359.400000       

       Fly Ash (component 3)(kg in a m^3 mixture)  \
count                                 1030.000000   
mean                                    54.187136   
std                                     63.996469   
min                                      0.000000   
25%                                      0.000000   
50%                                      0.000000   
75%                                    118.270000   
max                                    200.100000   

       Water  (component 4)(kg in a m^3 mixture)  \
count                                1030.000000   
mean                                  181.566359   
std                                    21.355567   
min                                   121.750000   
25%                                   164.900000   
50%                                   185.000000   
75%                                   192.000000   
max                                   247.000000   

       Superplasticizer (component 5)(kg in a m^3 mixture)  \
count                                        1030.000000     
mean                                            6.203112     
std                                             5.973492     
min                                             0.000000     
25%                                             0.000000     
50%                                             6.350000     
75%                                            10.160000     
max                                            32.200000     

       Coarse Aggregate  (component 6)(kg in a m^3 mixture)  \
count                                        1030.000000      
mean                                          972.918592      
std                                            77.753818      
min                                           801.000000      
25%                                           932.000000      
50%                                           968.000000      
75%                                          1029.400000      
max                                          1145.000000      

       Fine Aggregate (component 7)(kg in a m^3 mixture)    Age (day)  \
count                                        1030.000000  1030.000000   
mean                                          773.578883    45.662136   
std                                            80.175427    63.169912   
min                                           594.000000     1.000000   
25%                                           730.950000     7.000000   
50%                                           779.510000    28.000000   
75%                                           824.000000    56.000000   
max                                           992.600000   365.000000   

       Concrete compressive strength(MPa, megapascals)   
count                                       1030.000000  
mean                                          35.817836  
std                                           16.705679

In [5]:
X = np.array(df.drop(axis=1, columns=['Concrete compressive strength(MPa, megapascals) ']).values)
y = np.array(df['Concrete compressive strength(MPa, megapascals) '].values)

## Part 1 - Training linear regression model using normal equation method.

#### Adding a column of ones to represent the bias term

In [6]:
# X_old is the dataset which does not contain the column of ones, will be kept for later use
X_old = X 
X = np.hstack([np.ones((X.shape[0], 1)), X])

#### Splitting the dataset into train and test datasets

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

#### Implementing linear regression algorithm using normal equation method

In [8]:
theta = np.array(np.linalg.inv((X_train.T @ X_train)) @ X_train.T @ y_train)

#### Making predictions for train dataset

In [9]:
y_train_pred = X_train@(theta.T)

In [10]:
mse_train = mean_squared_error(y_train, y_train_pred)
r2_train = r2_score(y_train, y_train_pred)

print('Mean squared error of the model on the train dataset: {}'.format(mse_train))
print('r-squared value of the model on the train dataset: {}'.format(r2_train))

Mean squared error of the model on the train dataset: 107.25842311011503
r-squared value of the model on the train dataset: 0.6196006621677799


#### Making predictions for test dataset

In [11]:
y_test_pred = X_test@(theta.T)

In [12]:
mse_test = mean_squared_error(y_test, y_test_pred)
r2_test = r2_score(y_test, y_test_pred)

print('Mean squared error of the model on the test dataset: {}'.format(mse_test))
print('r-squared value of the model on the test dataset: {}'.format(r2_test))

Mean squared error of the model on the test dataset: 109.75614063735658
r-squared value of the model on the test dataset: 0.5943642890037111


For the previous assignment where linear regression was implemented using scikit-learn,    
(for random state = 42 while splitting train and test datasets)    

Mean squared error of the model on the train dataset: 107.25842311011506    
r-squared value of the model on the train dataset: 0.6196006621677799

Mean squared error of the model on the test dataset: 109.75614063734939    
r-squared value of the model on the test dataset: 0.5943642890037375

So, the model trained with the normal equation and the scikit-learn model are almost equally good.

## Part2 - L2 regularisation is introduced in the model.
Note: We will continue using the normal equation method to solve the linear regression problem    
   
Now, we will solve for different values of the regularisation parameter `lambda`

In [13]:
# Helper function for implementing linear regression with regularisation
def train_and_report(lamb = 0):

    # Solving the L2 regularised regression problem using normal equation
    theta = np.array(np.linalg.inv((X_train.T @ X_train) + lamb*np.identity(X_train.shape[1])) @ X_train.T @ y_train)

    y_train_pred = X_train@(theta.T)
    y_test_pred = X_test@(theta.T)

    mse_train = mean_squared_error(y_train, y_train_pred)
    r2_train = r2_score(y_train, y_train_pred)

    mse_test = mean_squared_error(y_test, y_test_pred)
    r2_test = r2_score(y_test, y_test_pred)

    print('For lambda value: {}'.format(lamb))
    print('Mean squared error of the model on the train dataset: {}'.format(mse_train))
    print('r-squared value of the model on the train dataset: {}'.format(r2_train))
    print('Mean squared error of the model on the test dataset: {}'.format(mse_test))
    print('r-squared value of the model on the test dataset: {}'.format(r2_test))

#### Training the regularised model for different values of lambda and reporting results

In [14]:
train_and_report(0.01)

For lambda value: 0.01
Mean squared error of the model on the train dataset: 107.26206768150276
r-squared value of the model on the train dataset: 0.6195877364459357
Mean squared error of the model on the test dataset: 109.59292614885034
r-squared value of the model on the test dataset: 0.5949674955733449


In [15]:
train_and_report(0.1)

For lambda value: 0.1
Mean squared error of the model on the train dataset: 107.38050914282948
r-squared value of the model on the train dataset: 0.6191676756977527
Mean squared error of the model on the test dataset: 108.89466307465511
r-squared value of the model on the test dataset: 0.5975481296674271


In [16]:
train_and_report(1)

For lambda value: 1
Mean squared error of the model on the train dataset: 107.70780493979476
r-squared value of the model on the train dataset: 0.6180068987551992
Mean squared error of the model on the test dataset: 108.28040562430594
r-squared value of the model on the test dataset: 0.5998182965679781


In [17]:
train_and_report(10)

For lambda value: 10
Mean squared error of the model on the train dataset: 107.80331053069902
r-squared value of the model on the train dataset: 0.6176681816410947
Mean squared error of the model on the test dataset: 108.17088388729358
r-squared value of the model on the test dataset: 0.6002230659722645


In [18]:
train_and_report(10000)

For lambda value: 10000
Mean squared error of the model on the train dataset: 108.14027174113436
r-squared value of the model on the train dataset: 0.6164731256482134
Mean squared error of the model on the test dataset: 107.66315179158745
r-squared value of the model on the test dataset: 0.6020995374702729


In [19]:
train_and_report(1e6)

For lambda value: 1000000.0
Mean squared error of the model on the train dataset: 134.24535553553258
r-squared value of the model on the train dataset: 0.5238896594597457
Mean squared error of the model on the test dataset: 136.27893280174166
r-squared value of the model on the test dataset: 0.49634160348714906


#### a. On training the model for different lambda values, it is observed that the model's performance on the test dataset gradually improves on increasing the value of lambda to a certain extent (order of 10<sup>4</sup>), after then the performance declines.

#### b. The model was earlier overfitting on the train dataset, as it tried to learn all the features and build a complex relation between them. Regularisation reduced the importance of certain features and forced the learning algorithm to build a less complex model, helping to generalise better.

#### If we increase the lambda value too much then it will lead to underfitting.

## Part 3 - Quadratic model instead of linear model

In [20]:
quad = PolynomialFeatures(2)
X_new = quad.fit_transform(X_old)
# Here we are using X_old because the quad object will anyway create a column of ones to represent the bias term, so no need of adding it beforehand

In [21]:
X_new.shape

(1030, 45)

#### a. We see that there are total 45 features (including the column of ones that represents the bias term)

#### b. Now solving again the linear regression problem using normal equation method

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X_new, y, test_size=0.3, random_state=42)

theta = np.array(np.linalg.inv((X_train.T @ X_train)) @ X_train.T @ y_train)

#### Making predictions and reporting results for train and test datasets

In [23]:
y_train_pred = X_train@(theta.T)
y_test_pred = X_test@(theta.T)

In [24]:
mse_train = mean_squared_error(y_train, y_train_pred)
r2_train = r2_score(y_train, y_train_pred)

print('Mean squared error of the model on the train dataset: {}'.format(mse_train))
print('r-squared value of the model on the train dataset: {}'.format(r2_train))

Mean squared error of the model on the train dataset: 51.599771287198465
r-squared value of the model on the train dataset: 0.8169978798793918


In [25]:
mse_test = mean_squared_error(y_test, y_test_pred)
r2_test = r2_score(y_test, y_test_pred)

print('Mean squared error of the model on the test dataset: {}'.format(mse_test))
print('r-squared value of the model on the test dataset: {}'.format(r2_test))

Mean squared error of the model on the test dataset: 60.041133169083835
r-squared value of the model on the test dataset: 0.7781005454397802


#### c. We see that the performance has improved significantly on adding quadratic features.

#### d. Number of features for cubic model.

In [26]:
X_old.shape

(1030, 8)

In [27]:
cubic = PolynomialFeatures(3)
X_cubic = cubic.fit_transform(X_old)
X_cubic.shape

(1030, 165)

#### We would need 165 features if we trained a cubic model.